drivers/interrupt_controller/intc_gicv3.c

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright (c) 2019 Broadcom
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <assert.h>
+#include <sw_isr_table.h>
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <drivers/interrupt_controller/gic.h>
+
+/* TODO: probe for each implemented rdist base */
+#define GIC_RDIST_BASE  DT_REG_ADDR_BY_IDX(DT_INST(0, arm_gic), 1)
+
+#define GICR_SGIBASE_OFFSET	0x10000
+
+/* Offsets from GICD base or GICR sgi_base */
+#define	GIC_CTRL		0x0
+#define	GIC_TYPER		0x4
+#define	GIC_IIDR		0x8
+#define	GIC_IGROUPR		0x80
+#define	GIC_ISENABLER		0x100
+#define	GIC_ICENABLER		0x180
+#define	GIC_ISPENDR		0x200
+#define	GIC_ICPENDR		0x280
+#define	GIC_ISACTIVER		0x300
+#define	GIC_ICACTIVER		0x380
+#define	GIC_IPRIORITYR		0x400
+#define	GIC_ITARGETSR		0x800
+#define	GIC_ICFGR		0xc00
+#define GIC_IGROUPMODR		0xd00
+#define	GIC_SGIR		0xf00
+
+#define	GICD_CTRL		GIC_DIST_BASE
+#define	GICR_CTRL		GIC_RDIST_BASE
+
+#define GICR_WAKER		0x14
+#define GIC_WAKER_PS		0x1
+#define GIC_WAKER_CA		0x2
+
+#define IGROUPR(base, n)	(base + GIC_IGROUPR + (n) * 4)
+#define ISENABLER(base, n)	(base + GIC_ISENABLER + (n) * 4)
+#define ICENABLER(base, n)	(base + GIC_ICENABLER + (n) * 4)
+#define ISPENDR(base, n)	(base + GIC_ISPENDR + (n) * 4)
+#define ICPENDR(base, n)	(base + GIC_ICPENDR + (n) * 4)
+#define IPRIORITYR(base, n)	(base + GIC_IPRIORITYR + n)
+#define ITARGETSR(base, n)	(base + GIC_ITARGETSR + (n) * 4)
+#define ICFGR(base, n)		(base + GIC_ICFGR + (n) * 4)
+#define IGROUPMODR(base, n)	(base + GIC_IGROUPMODR + (n) * 4)
+
+#define NUM_INTR_PER_REG	0x1F
+/* system register interface to GICv3 */
+#define ICC_IGRPEN1_EL1         S3_0_C12_C12_7
+#define ICC_SGI1R               S3_0_C12_C11_5
+#define ICC_SRE_EL1             S3_0_C12_C12_5
+#define ICC_SRE_EL2             S3_4_C12_C9_5
+#define ICC_SRE_EL3             S3_6_C12_C12_5
+#define ICC_CTLR_EL1            S3_0_C12_C12_4
+#define ICC_CTLR_EL3            S3_6_C12_C12_4
+#define ICC_PMR_EL1             S3_0_C4_C6_0
+#define ICC_RPR_EL1             S3_0_C12_C11_3
+#define ICC_IGRPEN1_EL3         S3_6_c12_c12_7
+#define ICC_IGRPEN0_EL1         S3_0_c12_c12_6
+#define ICC_HPPIR0_EL1          S3_0_c12_c8_2
+#define ICC_HPPIR1_EL1          S3_0_c12_c12_2
+#define ICC_IAR0_EL1            S3_0_c12_c8_0
+#define ICC_IAR1_EL1            S3_0_c12_c12_0
+#define ICC_EOIR0_EL1           S3_0_c12_c8_1
+#define ICC_EOIR1_EL1           S3_0_c12_c12_1
+#define ICC_SGI0R_EL1           S3_0_c12_c11_7
```

Comment:
Since these are cpu core system registers I am moving these register defines and their access macro to cpu.h  and moving to another PR.